internal/output/output.go

```diff
@@ -0,0 +1,69 @@
+package output
+
+import (
+	"strings"
+
+	"ldcli/internal/errors"
+)
+
+var ErrInvalidOutputKind = errors.NewError("output is invalid")
+
+// Outputter defines the different ways a command's response can be formatted based on
+// user input.
+type Outputter struct {
+	outputFn     PlaintextOutputFn
+	resource     resource
+	resourceJSON []byte
+}
+
+func (o Outputter) JSON() string {
+	return string(o.resourceJSON)
+}
+
+func (o Outputter) String() string {
+	return formatColl([]resource{o.resource}, o.outputFn)
+}
+
+// PlaintextOutputFn represents the various ways to output a resource or resources.
+type PlaintextOutputFn func(resource) string
+
+// OutputterFn is a factory to build the right outputter. By adding an layer of abstraction,
+// it lets us push back the error handling from where a caller provides the input to where
+// the caller builds the outputter.
+type OutputterFn interface {
+	New() (Outputter, error)
+}
+
+// resource is the subset of data we need to display a command's plain text response.
+type resource struct {
+	Key  string `json:"key"`
+	Name string `json:"name"`
+}
+
+// CmdOutput returns a command's response as a string formatted based on the user's requested type.
+func CmdOutput(outputKind string, outputter OutputterFn) (string, error) {
```

Comment:
This is the main function in this package. We use the `OutputterFn` factory so we can do error handling here instead of at the `CmdOutput` call site in the command handler.
```
// with factory fn
output, err := output.CmdOutput(
    viper.GetString(cliflags.OutputFlag),
    output.SingularOutput(response),
)
if err != nil {
    return err
}
```
vs
```
// without factory fn
outputter, err := output.SingularOutput(response)
if err != nil {
    return err
}
output, err := output.CmdOutput(
    viper.GetString(cliflags.OutputFlag),
    outputter
)
```